library/src/scala/IArray.scala

```diff
@@ -521,4 +601,575 @@ object IArray:
+    }
+
+    /** Builds a new array by applying a function to all elements of this array
+      * and using the elements of the resulting collections.
+      *
+      *  @param f      the function to apply to each element.
+      *  @tparam U     the element type of the returned array.
+      *  @return       a new array resulting from applying the given collection-valued function
+      *                `f` to each element of this array and concatenating the results.
+      */
+    def flatMap[U: ClassTag](f: T => IterableOnce[U]): IArray[U] = {
+      val b = IArrayBuilder.make[U]
+      var i = 0
+      while(i < xs.length) {
+        val x = xs(i)
+        if(p(x)) b ++= f(xs(i))
+        i += 1
+      }
+      b.result()
+    }
+
+    def flatMap[BS, U](f: T => BS)(using asIterable: BS => Iterable[U], m: ClassTag[U]): IArray[U] =
+      flatMap[U](x => asIterable(f(x)))
+
+    /** Creates a new non-strict filter which combines this filter with the given predicate. */
+    def withFilter(q: T => Boolean): WithFilter[T] = new WithFilter[T](a => p(a) && q(a), xs)
+
+  end WithFilter
+
+  /** A builder class for immutable arrays.
+  *
+  *  @tparam T    the type of the elements for the builder.
+  */
+  @SerialVersionUID(3L)
+  sealed abstract class IArrayBuilder[T]
+    extends mutable.ReusableBuilder[T, IArray[T]]
+      with Serializable {
+    protected[this] var capacity: Int = 0
+    protected[this] def elems: Array[T]
+    protected var size: Int = 0
+
+    def length: Int = size
+
+    override def knownSize: Int = size
+
+    protected[this] final def ensureSize(size: Int): Unit = {
+      if (capacity < size || capacity == 0) {
+        var newsize = if (capacity == 0) 16 else capacity * 2
+        while (newsize < size) newsize *= 2
+        resize(newsize)
+      }
+    }
+
+    override final def sizeHint(size: Int): Unit =
+      if (capacity < size) resize(size)
+
+    def clear(): Unit = size = 0
+
+    protected[this] def resize(size: Int): Unit
+
+    /** Add all elements of an array */
+    def addAll(xs: IArray[T]): this.type = addAll(xs, 0, xs.length)
+
+    /** Add a slice of an array */
+    def addAll(xs: IArray[T], offset: Int, length: Int): this.type = {
+      ensureSize(this.size + length)
+      Array.copy(xs.asInstanceOf[Array[T]], offset, elems, this.size, length)
+      size += length
+      this
+    }
+
+    override def addAll(xs: IterableOnce[T]): this.type = {
+      val k = xs.knownSize
+      if (k > 0) {
+        ensureSize(this.size + k)
+        // IterableOnce.copyElemsToArray(xs, elems, this.size)
+        // manually inlined as
+        xs match {
+          case src: Iterable[T] => src.copyToArray[T](elems, this.size, Int.MaxValue)
+          case src              => src.iterator.copyToArray[T](elems, this.size, Int.MaxValue)
+        }
+        size += k
+      } else if (k < 0) super.addAll(xs)
+      this
+    }
+  }
+
+  /** A companion object for array builders.
+  */
+  object IArrayBuilder {
```

Comment:
There should be a way to re-use array builders here. I believe there is an operation where you create a builder by using an existing builder and adding a post-processing step. In this case the post-processing step would be simply `asInstanceOf[IArray]`. Can we use that one?

Reply:
We couls still have specialized versions for each primitive type, but we would not need to manually re-implement all the methods.

Reply:
Indeed! This should probably be as easy as
```scala
@inline def makeIArrayBuilder[T: ClassTag]: Builder[T, IArray[T]] =
  mutable.ArrayBuilder.make[T].mapResult(r => r: IArray[T])
```

Reply:
I reimplemeted `IArrayBuilder` and moved it to `scala.collection.mutable`.